_episodes/07-raster-calculations.md

```diff
@@ -3,303 +3,367 @@ title: "Raster Calculations in Python"
 > 2. Plot a histogram with 50 bins instead of 8. What do you notice that wasn't clear before?
-> 3. Plot the `canopy_HARV` raster using breaks that make sense for the data. Include an appropriate color palette for the data, plot title and no axes ticks / labels.
-> 
+> 3. Plot the `ndvi` raster using breaks that make sense for the data.
+>
 > > ## Answers
-> > 
-> > 1) Recall, if there were nodata values in our raster like `-9999.0`, 
+> >
+> > 1) Recall, if there were nodata values in our raster,
 > > we would need to filter them out with `.where()`.
-> > ```python
-> > canopy_HARV.min().values
-> > canopy_HARV.max().values
-> > ```
 > > ~~~
-> > array(-1.)
-> > array(38.16998291)
+> > print(ndvi.min().values)
+> > print(ndvi.max().values)
+> > ~~~
+> > {: .language-python}
+> > ~~~
+> > -0.99864775
+> > 0.9995788
 > > ~~~
 > > {: .output}
-> > 2) Increasing the number of bins gives us a much clearer view of the distribution.
-> > ```python
-canopy_HARV.plot.hist(bins=50)
-> > ```
-> > ![](../fig/07-HARV-CHM-histo-50bins-03.png) 
+> > 2) Increasing the number of bins gives us a much clearer view of the distribution. Also, there are almost no NDVI
+> > values larger than ~0.9.
+> > ~~~
+> > ndvi.plot.hist(bins=50)
+> > ~~~
+> > {: .language-python}
+> > ![](../fig/E09-06-NDVI-hist-bins.png)
+> > 3) We can discretize the color bar by specifying the intervals via the `level` argument to `plot()`.
+> > Suppose we want to bin our data in the following intervals:
+> > * $-1 \le NDVI \lt 0$ for water;
+> > * $0 \le NDVI \lt 0.2$ for no vegetation;
+> > * $0.2 \le NDVI \lt 0.7$ for sparse vegetation;
+> > * $0.7 \le NDVI \lt 1$ for dense vegetation.
+> >
+> > ~~~
+> > class_bins = (-1, 0., 0.2, 0.7, 1)
+> > ndvi.plot(levels=class_bins)
+> > ~~~
+> > {: .language-python}
+> > ![](../fig/E09-07-NDVI-map-binned.png)
 > {: .solution}
 {: .challenge}
 
 ## Classifying Continuous Rasters in Python
 
-Now that we have a sense of the distribution of our canopy height raster, we
+Now that we have a sense of the distribution of our NDVI raster, we
 can reduce the complexity of our map by classifying it. Classification involves
 assigning each pixel in the raster to a class based on its value. In Python, we
 can accomplish this using the `numpy.digitize` function.
 
-First, we define canopy height classes based on a list of heights:
-`[canopy_HARV.min().values, 2, 10, 20, np.inf]`. When bins are ordered from
+First, we define NDVI classes based on a list of values, as defined in the last exercise:
+`[-1, 0., 0.2, 0.7, 1]`. When bins are ordered from
 low to high, as here, `numpy.digitize` assigns classes like so:
 
-![Canopy height classes](../fig/07-CHM-classification-bins-01.png)
+![](../fig/E09-08-NDVI-classes.jpg)
 
 Source: Image created for this lesson ([license](../LICENSE.md))
 {: .text-center}
 
 Note that, by default, each class includes the left but not right bound.
```

Comment:
What happens for a value of NDVI = 1, assuming that an NDVI of 1 is possible?

Reply:
NDVI values of 1 are actually possible, but they do not appear with the current datasets. I have added a small comment trying to address the confusion that might arise from the fact that right-open intervals are used in the classification, but didn't want to enter too much in the discussion since anyway NDVI values are included in the open range (-1;1). Do you think this is clear right now? 

I have instead added one step to get rid of the NaN values (actually, value) in the NDVI, which were responsible for a fifth class being present as result of the binning. `ndvi_classified` now includes only classes from 1 to 4.